nose/util.py

```diff
@@ -449,16 +449,35 @@ def try_run(obj, names):
             if type(obj) == types.ModuleType:
                 # py.test compatibility
                 if isinstance(func, types.FunctionType):
-                    args, varargs, varkw, defaults = \
-                        inspect.getargspec(func)
+                    try:
+                        sig = inspect.signature(func)
+                        bl = set([inspect.Parameter.VAR_KEYWORD,
+                                  inspect.Parameter.VAR_POSITIONAL,
+                                  inspect.Parameter.KEYWORD_ONLY])
+                        args = [k for k, v in sig.parameters.items()
+                                if v.kind not in bl]
+                    except AttributeError:
+                        args, varargs, varkw, defaults = \
+                          inspect.getargspec(func)
+
                 else:
                     # Not a function. If it's callable, call it anyway
                     if hasattr(func, '__call__') and not inspect.ismethod(func):
                         func = func.__call__
                     try:
-                        args, varargs, varkw, defaults = \
-                            inspect.getargspec(func)
-                        args.pop(0) # pop the self off
+                        try:
+                            sig = inspect.signature(func)
+                            bl = set([inspect.Parameter.VAR_KEYWORD,
+                                      inspect.Parameter.VAR_POSITIONAL,
+                                      inspect.Parameter.KEYWORD_ONLY])
+                            args = [k for k, v in sig.parameters.items()
+                                    if v.kind not in bl]
+
+                        except AttributeError:
+                            args, varargs, varkw, defaults = \
+                                inspect.getargspec(func)
+                            # signature never returns it
```

Comment:
Are you sure about this?

```
>>> import inspect
>>> class foo:
...   def x(self):
...     pass
... 
>>> inspect.signature(foo.x)
<Signature (self)>
>>> list(inspect.signature(foo.x).parameters.items())
[('self', <Parameter "self">)]
```

Reply:
Ah thx will remove that comment, I see this is the old code path and the old code did not have this comment, I think I added this comment before adding the new code path